File: java/0020.valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        LinkedList<Character> stack = new LinkedList<Character>();
        for(int i=0;i < s.length(); ++i) {
            Character c = s.charAt(i);
            if(c=='('||c=='['||c=='{'){
                stack.addFirst(c);
            }else if(stack.isEmpty()){
                return false;
            } else if (c == ')' && stack.peekFirst() == '('){
                stack.pollFirst();
            }else if (c == ']' && stack.peekFirst() == '['){
                stack.pollFirst();
            }else if (c == '}' && stack.peekFirst() == '{'){
                stack.pollFirst();
            }else{
                return false;
            }
        }
        if(stack.peekFirst() == null){
            return true;
        }else{
            return false;
        }
    }
}

File: python/0020. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for i in s:
            match i:
                case ')':
                    if len(stack) and stack[-1]=='(':
                        stack.pop()
                    else:
                        return False
                case '}':
                    if len(stack) and stack[-1]=='{':
                        stack.pop()
                    else:
                        return False
                case ']':
                    if len(stack) and stack[-1]=='[':
                        stack.pop()
                    else:
                        return False
                case _:
                    stack.append(i)

        if len(stack):
            return False
        return True
